context.go

```diff
@@ -6,53 +6,142 @@ import (
 	"flag"
 	"fmt"
 	"strings"
+	"time"
 )
 
-// Context is a type that is passed through to
-// each Handler action in a cli application. Context
+// Context is an interface that is passed through to
+// each Handler action in a cli application. It
 // can be used to retrieve context-specific args and
 // parsed command-line options.
-type Context struct {
-	context.Context
-	App           *App
-	Command       *Command
+type Context interface {
+	// Functions to be able to manage Context type
+	App() *App
+	SetCommand(command *Command)
+	Command() *Command
+	Context() context.Context
+	ParentContext() Context
+	SetShellComplete(shellComplete bool)
+	ShellComplete() bool
+	SetFlagSet(set *flag.FlagSet)
+	FlagSet() *flag.FlagSet
+	NumFlags() int
+	Set(name, value string) error
+	IsSet(name string) bool
+	LocalFlagNames() []string
+	FlagNames() []string
+	Lineage() []Context
+	Value(name string) interface{}
+	Args() Args
+	NArg() int
+	// Functions for each flag type supported by the cli
+	Timestamp(name string) *time.Time
+	Generic(name string) interface{}
+	Bool(name string) bool
+	Float64(name string) float64
+	String(name string) string
+	Int64Slice(name string) []int64
+	Uint64(name string) uint64
+	StringSlice(name string) []string
+	Int64(name string) int64
+	Float64Slice(name string) []float64
+	Duration(name string) time.Duration
+	IntSlice(name string) []int
+	Int(name string) int
+	Path(name string) string
+	Uint(name string) uint
+}
+
+type cliContext struct {
+	context       context.Context
+	app           *App
+	command       *Command
 	shellComplete bool
 	flagSet       *flag.FlagSet
-	parentContext *Context
+	parentContext Context
 }
 
 // NewContext creates a new context. For use in when invoking an App or Command action.
-func NewContext(app *App, set *flag.FlagSet, parentCtx *Context) *Context {
-	c := &Context{App: app, flagSet: set, parentContext: parentCtx}
+func NewContext(app *App, set *flag.FlagSet, parentCtx Context) Context {
```

Comment:
One of the alternatives to this PR as a whole would be to revamp the `NewContext` function to actually be useful.

The problem it has now is that the interface makes it impossible to create a useful `cli.Context` outside of the library. Accepting a `*flag.FlagSet`, which is from the standard library, makes it look generic, but the issue is that the flag set creation logic we use internally is non-trivial and relies on unexported fields. Specifically, we:
- Use a private function to check if we're doing shell completion
- Use a private function to create a new flag set
- Use a private function to parse flags iteratively
- Use a private function to normalize flags
- Handle errors in a way that sometimes uses the context even if creating it has failed
- Modify private fields on the context immediately after its used

On top of this, the logic is similar but slightly different between doing those steps for an app versus a command.

I could see a much smaller breaking change solving the same problem by doing something like this:
```go
func (a *App) NewContext(args []string, parent Context) Context {
```

Switching from a `*flag.FlagSet` to `[]string` means we can continue to own the complicated flag parsing logic, while giving users a way to create new contexts in tests as simply as passing flags. It also means the `cli.Context` in tests doesn't need to be mocked, making tests slightly closer to what real code would be doing and potentially making test boilerplate smaller.

I'm not sure if that would be better than what this PR does, or what the implications of that change would be in terms of broader refactors, but I wanted to float that idea before we decide we're definitely going in this direction.

Reply:
This is interesting. Making `NewContext` a part of `App` will lead to a large number of changes in the code and the tests. I would be willing to do it if we are left with no other option. I have been reading around to help me solve this and I think we can take inspiration from the standard library and have a number of functions like:

```go
func NewContext(context context.Context) Context
func WithParent(parent Context) Context
func WithApp(app *App) Context
func WithFlagset(set *flag.FlagSet) Context
```

Such kind of functions are available in `html/template` `text/template` and `context` standard library. I think these would allow us to be much flexible with how the context is created.

What do you think?

Reply:
I'm not sure this is a good fit here.

The biggest goals I would prioritize for a revamped `NewContext` function would be to make it as simple as possible to generate a valid `Context`, and to make it difficult or impossible to generate an invalid `Context`.

For the first goal, the reason I suggested the signature just taking a list of args is because that's an interface that users already are familiar with, and it's data they have access to. `app.Run` takes a list of args. The CLI itself effectively takes a list of args. `os.Args` gives easy access in their main program, and so on.

The challenge with chaining is that you get a `Context` back at every step. Is a `Context` valid without an app? Is it valid without any flag set? I believe we'd end up in a similar situation to now, where it would be easy to generate an invalid `Context`, and not necessarily simpler to create a valid one.

Reply:
The real question that I am thinking about is whether the `NewContext` function should even be exposed. I have not yet come across a scenario where I needed to create an instance of `cli.Context`. 

Also to answer your question, the `NewContext` function can be modified to set sane defaults.

Reply:
> I have not yet come across a scenario where I needed to create an instance of cli.Context.

If this is the case, what's the benefit in making `cli.Context` an interface? 

Interfaces are are less useful than structs in nearly every way. Compared to a struct, using an interface is more verbose because you have to access fields by function call rather than directly, harder to instantiate because you need a minimum of one additional type implemented to be able to use them, less performant because you're forcing an allocation, less clear because "go-to definition" in the editor now only reaches a signature rather than an implementation, and so on.

The benefit, though, is that you aren't tied to a single implementation. 

If you aren't reaching to create a `Context`, then I don't think this change brings us value in terms of public API. The big benefit of being able to provide an alternate implementation is irrelevant—there's no need to provide any implementation at all.

The reason I'm suggesting changing `NewContext` instead of making `cli.Context` an interface is because the specific goal I had in mind when suggesting making `Context` into an interface was to make a `Context` easier to create. For me, this is for unit testing.

Here is [an example of code where the only things I've needed out of a context are `NArg` and `IsSet`](https://github.com/rliebz/tusk/blob/f27bf1c90703ca30ea931bf583b56245d79e2adb/appcli/completion.go). It's awkwardly implemented because it was not possible with the current interface of `NewContext` for me to create a `Context` that was close enough to the real thing that it would be useful in tests. This awkward implementation does [let me create a context for myself](https://github.com/rliebz/tusk/blob/master/appcli/completion_test.go#L420-L437) that I can [instantiate in unit tests without having to manually parse arguments myself](https://github.com/rliebz/tusk/blob/f27bf1c90703ca30ea931bf583b56245d79e2adb/appcli/completion_test.go#L261-L264), because the logic to do so is [a private implementation detail of `urfave/cli` and incredibly non-trivial](https://github.com/urfave/cli/blob/fed41d6ef81daa62b81c858105cde642a31cd976/app.go#L217-L233). This PR would _definitely_ make some of those code snippets more straightforward and easier to test, but before we tie ourselves to this breaking change in particular, I want to explore if there are better ways to solve the same problems.

> Making NewContext a part of App will lead to a large number of changes in the code and the tests. 

The other point I want to make here is that I'm less concerned about the size of the change within `urfave/cli` and more concerned about the size of the change to dependent applications. I can spend a few hours on a weekend patching all the unit tests for `urfave/cli` if we make an internally-facing breaking change, no big deal. If we make an externally-facing breaking change, I'll still spend a few hours on a weekend patching some other applications, but so will everyone else who uses this project.